samza-core/src/main/java/org/apache/samza/execution/ExecutionPlanner.java

```diff
@@ -0,0 +1,337 @@
+      // set the partitions of a stream to its StreamEdge
+      streamToMetadata.forEach((stream, data) -> {
+          int partitions = data.getSystemStreamPartitionMetadata().size();
+          streamToStreamEdge.get(stream).setPartitions(partitions);
+          log.debug("Partition count is {} for stream {}", partitions, stream);
+        });
+    }
+  }
+
+  /**
+   * Calculate the partitions for the input streams of join operators
+   */
+  /* package private */ static void calculateJoinInputPartitions(StreamGraph streamGraph, ProcessorGraph processorGraph) {
+    // mapping from a source stream to all join specs reachable from it
+    Multimap<OperatorSpec, StreamEdge> joinSpecToStreamEdges = HashMultimap.create();
+    // reverse mapping of the above
+    Multimap<StreamEdge, OperatorSpec> streamEdgeToJoinSpecs = HashMultimap.create();
+    // Mapping from the output stream to the join spec. Since StreamGraph creates two partial join operators for a join and they
+    // will have the same output stream, this mapping is used to choose one of them as the unique join spec representing this join
+    // (who register first in the map wins).
+    Map<MessageStream, OperatorSpec> outputStreamToJoinSpec = new HashMap<>();
+    // A queue of joins with known input partitions
+    Queue<OperatorSpec> joinQ = new LinkedList<>();
+    // The visited set keeps track of the join specs that have been already inserted in the queue before
+    Set<OperatorSpec> visited = new HashSet<>();
+
+    streamGraph.getInStreams().entrySet().forEach(entry -> {
+        StreamEdge streamEdge = processorGraph.getOrCreateEdge(entry.getKey());
+        // Traverses the StreamGraph to find and update mappings for all Joins reachable from this input StreamEdge
+        findReachableJoins(entry.getValue(), streamEdge, joinSpecToStreamEdges, streamEdgeToJoinSpecs,
+            outputStreamToJoinSpec, joinQ, visited);
+      });
+
+    // At this point, joinQ contains joinSpecs where at least one of the input stream edge partitions is known.
+    while (!joinQ.isEmpty()) {
+      OperatorSpec join = joinQ.poll();
+      int partitions = -1;
+      // loop through the input streams to the join and find the partition count
+      for (StreamEdge edge : joinSpecToStreamEdges.get(join)) {
+        int edgePartitions = edge.getPartitions();
+        if (edgePartitions != -1) {
+          if (partitions == -1) {
+            //if the partition is not assigned
+            partitions = edgePartitions;
+          } else if (partitions != edgePartitions) {
+            throw  new SamzaException(String.format(
+                "Unable to resolve input partitions of stream %s for join. Expected: %d, Actual: %d",
+                edge.getFormattedSystemStream(), partitions, edgePartitions));
+          }
+        }
+      }
+      // assign the partition count
+      for (StreamEdge edge : joinSpecToStreamEdges.get(join)) {
+        if (edge.getPartitions() <= 0) {
```

Comment:
What conditions would cause an `edge.getPartitions() == -1` at this point? Will be super-helpful to document this.